### ISSUE TITLE
recipe00's reformA.json has comments that make it invalid

### ISSUE THREAD

**Issue by MaxGhenis** [PSLmodels/Tax-Calculator#2259]
https://pslmodels.github.io/Tax-Calculator/recipe00.py.html points to https://pslmodels.github.io/Tax-Calculator/reformA.json, which has comments in between the lines. Loading it produced this error:

```
reform_filename = './ingredients/reformA.json'
# This points to the same:
# reform_filename = 'https://pslmodels.github.io/Tax-Calculator/reformA.json'
params = tc.Calculator.read_json_param_objects(reform_filename, None)
# ... Error ...
ValueError: Text below contains invalid JSON:
Expecting value: line 1 column 1 (char 0)
Above location of the first error may be approximate.
The invalid JSON text is between the lines:
XXXX----.----1----.----2----.----3----.----4----.----5----.----6----.----7
0001<title>reformA.json</title><pre>
0002 
0003 
0004{
0005    "policy": {
0006	 
0007	"_II_em": {"2020": [1000]},
0008	 
0009	 
0010	"_II_rt5": {"2020": [0.36]},
0011	"_II_rt6": {"2020": [0.39]},
0012	"_II_rt7": {"2020": [0.41]},
0013	 
0014	"_PT_rt5": {"2020": [0.36]},
0015	"_PT_rt6": {"2020": [0.39]},
0016	"_PT_rt7": {"2020": [0.41]}
0017    }
0018}
0019</pre>
0020
XXXX----.----1----.----2----.----3----.----4----.----5----.----6----.----7
```

Removing the comments in between made it run OK. The comments at the top are safely ignored.

**Comment by martinholmer** [comment]
@MaxGhenis,  Thanks for sharing your experience with cookbook recipe00 in [this comment](https://github.com/PSLmodels/Tax-Calculator/issues/2259#issue-421687320).  I'm sorry you're having difficulty following the recipe.

Let me try to explain.  You start out saying this:
> https://pslmodels.github.io/Tax-Calculator/recipe00.py.html points to https://pslmodels.github.io/Tax-Calculator/reformA.json, which ...

Actually, there is no file named https://pslmodels.github.io/Tax-Calculator/reformA.json, but there is a file named https://pslmodels.github.io/Tax-Calculator/reformA.json.html.  The cookbook is an HTML document so all its files end in `.html` and contain very simple but valid HTML code.  What recipe00 "points to" is a file named `ingredients/reformA.json` which is not an HTML file but a JSON file.  It is true that this JSON file contains some non-standard `//` comments, but the `Calculator.read_json_param_objects` method strips those comments out before trying to decode the JSON text.  The problem you experienced has nothing to do with the embedded comments, but rather stems from you providing the wrong argument to the `Calculator.read_json_param_objects` method.  I'm concluding that because of the error message you said your received, which I reproduce here:
```
ValueError: Text below contains invalid JSON:
Expecting value: line 1 column 1 (char 0)
Above location of the first error may be approximate.
The invalid JSON text is between the lines:
XXXX----.----1----.----2----.----3----.----4----.----5----.----6----.----7
0001<title>reformA.json</title><pre>
0002 
0003 
0004{
0005    "policy": {
0006	 
0007	"_II_em": {"2020": [1000]},
0008	 
0009	 
0010	"_II_rt5": {"2020": [0.36]},
0011	"_II_rt6": {"2020": [0.39]},
0012	"_II_rt7": {"2020": [0.41]},
0013	 
0014	"_PT_rt5": {"2020": [0.36]},
0015	"_PT_rt6": {"2020": [0.39]},
0016	"_PT_rt7": {"2020": [0.41]}
0017    }
0018}
0019</pre>
0020
XXXX----.----1----.----2----.----3----.----4----.----5----.----6----.----7
```
Notice the the JSON text it is trying to decode has the embedded comments removed, so those can't be the source of the problem you experienced.  The second line of the error message locates the problem on line 1 column 1, which contains this line: `<title>reformA.json</title><pre>`.  This is pretty hard evidence that you tried to read with the `Calculator.read_json_param_objects` method an HTML file rather than a JSON file.

So, the real problem was that the `Calculator.read_json_param_objects` method didn't stop you as soon as it determined that the filename (or string beginning with `http`) you supplied did not end with `.json`.  I've added that error checking to the `Calculator.read_json_param_objects` method in pull request #2261.

Again, thanks for sharing your experience using the Tax-Calculator cookbook.  I hope the "guard rails" I added in #2261 will provide more helpful error messages.

I'm going to close this issue (because my view is that it has been resolved by pull request #2261) and close your pull request #2260 because it focuses on a "problem" that you did not experience.

If you have any other questions or problems following the recipes in the Tax-Calculator cookbook, please raise another issue.

**Comment by MaxGhenis** [comment]
I linked to https://pslmodels.github.io/Tax-Calculator/reformA.json for reference only. When running the example I downloaded the json file locally and got the error above, which would not have occurred if it couldn't read the file at all (the error message includes parts of the json file). You can see in my code that `reform_filename` is a local file, so validating that the path ends in `.json` will not resolve this.

Are you able to run the example?

**Comment by martinholmer** [comment]
@MaxGhenis asked:
> Are you able to run the example?

No, I'm totally confused about what you did?

**Comment by MaxGhenis** [comment]
I see what happened. I called `wget https://pslmodels.github.io/Tax-Calculator/reformA.json` from the `ingredients` folder, which I thought would be a usable json file (it's also described as `reformA.json` when linked in the cookbook). Instead, it has the same contents as https://pslmodels.github.io/Tax-Calculator/reformA.json.html, including the html header, and this broke `Calculator.read_json_param_objects`. I thought the problem was the comments, and fixed those, but must have also removed the header inadvertently, which was the true fix.

Based on this, it seems that the intended way of running this recipe is:
1. Copy the contents of https://pslmodels.github.io/Tax-Calculator/reformA.json.html from the webpage into a text editor, and save that as `reformA.json` in an `ingredients` subfolder.
2. Copy the contents of https://pslmodels.github.io/Tax-Calculator/recipe00.py.html from the webpage into a text editor, and save that as `recipe00.py` (`wget` will produce the same problem for the `py` file).
3. Run `python recipe00.py > recipe00.out`.

I think these changes would be simpler to users (especially important if this intends to be their very first usage of taxcalc), in order of significance:
1. Replace the `.json.html` files with `.json` files, and point instead to the raw versions hosted on GitHub (like [this](https://raw.githubusercontent.com/MaxGhenis/taxcalc-notebooks/master/recipes/ingredients/reformA.json)). For example, these are stored for all files in the `reforms` folder, e.g. https://raw.githubusercontent.com/PSLmodels/Tax-Calculator/master/taxcalc/reforms/2017_law.json. At this point, a user can use `wget` to pull down a file without copy-pasting or removing headers.
2. Adapt the code to pull the reform file from the web, to remove the need for the `ingredients` folder and pulling the reform file.
3. Put it all in a self-contained notebook that "just works" (only requires `taxcalc` to be installed). I've done this at https://github.com/MaxGhenis/taxcalc-notebooks/blob/master/recipes/recipe00.ipynb.

**Comment by martinholmer** [comment]
@MaxGhenis said:
> I see what happened. I called `wget https://pslmodels.github.io/Tax-Calculator/reformA.json` from the `ingredients` folder, ...

Let me stop you right there.  The URL of the `cookbook/ingredients/reformA.json` file is this:
```
https://pslmodels.github.io/Tax-Calculator/cookbook/ingredients/reformA.json
```
If you fetch the only correct JSON file for reformA then I think all will work smoothly.

Here's what I get when I download the **correct** file:
```
iMac:PSL mrh$ curl https://pslmodels.github.io/Tax-Calculator/cookbook/ingredients/reformA.json
// Raise personal exemption and raise regular tax rates in top three brackets
// from (0.32, 0.35, 0.37) to (0.36, 0.39, 0.41) beginning in 2020.
{
    "policy": {
	// raise personal exemption amount and index in subsequent years
	"_II_em": {"2020": [1000]},
	// raise tax rates in the top three brackets
	// ... raise non-AMT rates on non-pass-through income
	"_II_rt5": {"2020": [0.36]},
	"_II_rt6": {"2020": [0.39]},
	"_II_rt7": {"2020": [0.41]},
	// ... raise non-AMT rates on pass-through income
	"_PT_rt5": {"2020": [0.36]},
	"_PT_rt6": {"2020": [0.39]},
	"_PT_rt7": {"2020": [0.41]}
    }
}
iMac:PSL mrh$ 
```
Notice there's no HTML tags in the downloaded file.

Here's the directory listing for the `cookbook/ingredients` directory:

<img width="1001" alt="Screen Shot 2019-03-16 at 4 34 19 PM" src="https://user-images.githubusercontent.com/12170745/54481435-eb88e180-480a-11e9-907e-103161e50f1f.png">

**Comment by MaxGhenis** [comment]
Odd, maybe `curl` and `wget` operate differently between Mac and Linux? I get the html tag using both, e.g. `wget`:

```
(xenial)maxghenis@localhost:~/taxcalc-notebooks/recipes$ wget https://pslmodels.github.io/Tax-Calculator/reformA.json
(xenial)maxghenis@localhost:~/taxcalc-notebooks/recipes$ head -1 reformA.json
<title>reformA.json</title><pre>
```

In any case is there a need to maintain these duplicate files, when the cookbook can link to https://raw.githubusercontent.com/PSLmodels/Tax-Calculator/master/docs/cookbook/ingredients/reformA.json directly? e.g. https://github.com/PSLmodels/Tax-Calculator/commit/7847010b28bbd4faad08276f3067ba8f2a986f4e required modifying both in parallel, a step that can be avoided.

**Comment by martinholmer** [comment]
@MaxGhenis said:
> Odd, maybe curl and wget operate differently between Mac and Linux? I get the html tag using both, e.g. wget:
>
>```
>(xenial)maxghenis@localhost:~/taxcalc-notebooks/recipes$ wget https://pslmodels.github.io/Tax-Calculator/reformA.json
>(xenial)maxghenis@localhost:~/taxcalc-notebooks/recipes$ head -1 reformA.json
><title>reformA.json</title><pre>
>```

No, I think they operate identically.
What you did above shows you didn't read carefully what I said earlier.  You're still pointing `wget` to an HTML file instead of a JSON file, and so you are are predictably getting a file that contains HTML tags.

**Comment by MaxGhenis** [comment]
OK I see, yes when pointed to the `cookbook/ingredients` file there's no HTML tag.

The core point remains: the `cookbook/ingredients` file isn't linked from the cookbook, so why would users be expected to know to download that? I don't think it's crazy for a new user to do what I did, which was to (a) click on the policy reform linked from the cookbook, (b) see that it works when removing `.html` from the URL, and (c) download that file.

What's the value in linking to this page that could confuse new users? Is the three-step process I outlined in https://github.com/PSLmodels/Tax-Calculator/issues/2259#issuecomment-473581235, with all the copy-pasting, really ideal for new users? It's not even what experienced users should be doing for real analysis IMO.
